syntax="proto3";
package frp;
// 编译命令: protoc message.proto --cpp_out=./

message Msg {
    MsgType type = 1;
    LoginReq loginReq = 2;
    LoginRsp loginRsp = 3;
    AddConnReq addConnReq = 4;
    AddConnRsp addConnRsp = 5;
}

enum MsgType {
    MSGTYPE_UNKNOWN = 0;
    MSGTYPE_LOGIN_REQ = 1;
    MSGTYPE_LOGIN_RSP = 2;
    MSGTYPE_ADD_CONN_REQ = 3;
    MSGTYPE_ADD_CONN_RSP = 4;
}

message LoginReq {
    // 携带配置告诉对端需要开启什么端口
    LocalConf localConf = 1;
}

message LoginRsp {
    int32 ret_code = 1;
    string err_msg = 2;
    // 携带cliennt_id，用以识别
    string client_id = 3;
}

// 公网服务器告诉客户端，需要新增一条tcp链接用以转发
// 内网服务器告诉公网服务器，这是新增链接请求
message AddConnReq {
    string conn_id = 1;
}

// 客户端发起链接后告诉公网服务器可以了
// 不过这里设计其实不够鲁棒，可以慢慢优化
message AddConnRsp {
    int32 ret_code = 1;
    string err_msg = 2;
    string conn_id = 3;
}
 
message Config {
    // 公网服务器
    ServerConf serverConf = 1;
    // 本地转发端口
    repeated LocalConf localConfs = 2;
}

message LocalConf {
    Protocol protocol = 1;
    string ip = 2;
    uint32 port = 3;
    uint32 openPort = 4;
}

enum Protocol {
    TCP = 0;
}

message ServerConf {
    string ip = 1;
    uint32 port = 2;
}